### PR TITLE
Re-factor format lint

### PR DESCRIPTION
**Pull request by lzutao**
cc #4432

changelog: none

### CODE REVIEW

**Comment by jeremystucki** [comment]
It seems like you need to update the ui tests in `ui/unicode.rs`

**Comment by phansch** [comment]
Retriggering the build

**Comment by phansch** [comment]
@lzutao looks like only some dogfood tests need to be fixed: https://travis-ci.com/rust-lang/rust-clippy/jobs/227661145#L1079

```
error: this constant should be before the previous constant due to lexical ordering
 --> clippy_lints/src/utils/paths.rs:31:1
   |
31 | pub const FMT_ARGUMENTS_NEW_V1: [&str; 4] = ["core", "fmt", "Arguments", "new_v1"];
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: `-D clippy::clippy-lints-internal` implied by `-D clippy::internal`
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#clippy_lints_internal
error: this constant should be before the previous constant due to lexical ordering

  --> clippy_lints/src/utils/paths.rs:31:1
   |
31 | pub const FMT_ARGUMENTS_NEW_V1: [&str; 4] = ["core", "fmt", "Arguments", "new_v1"];
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: `-D clippy::clippy-lints-internal` implied by `-D clippy::internal`
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#clippy_lints_internal
```

r=me with tests passing

**Comment by lzutao** [comment]
CI is green.

File: tests/ui/format.stderr [hunk 1]
```diff
@@ -19,52 +19,67 @@ LL |     format!("{{}} abc {{}}");
...
    |
 LL |     format!("{}", x.display().to_string());
-   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: `to_string()` is enough: `x.display().to_string();`
+   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: consider using .to_string(): `x.display().to_string();`
+
+error: useless use of `format!`
+  --> $DIR/format.rs:69:18
+   |
+LL |     let _ = Some(format!("{}", a + "bar"));
+   |                  ^^^^^^^^^^^^^^^^^^^^^^^^ help: consider using .to_string(): `a + "bar"`
```
Comment (mati865) [inline tests/ui/format.stderr]: Seems this should be not linted?

Comment (lzutao) [reply tests/ui/format.stderr]: Yes, it should. I forgot to removing the misleading comments.

**Pull request merged**